qiskit/algorithms/state_fidelities/base_state_fidelity.py

```diff
@@ -0,0 +1,299 @@
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Base state fidelity interface
+"""
+
+from __future__ import annotations
+from abc import ABC, abstractmethod
+from collections.abc import Sequence, Mapping
+import numpy as np
+
+from qiskit import QuantumCircuit
+from qiskit.algorithms import AlgorithmJob
+from qiskit.circuit import ParameterVector
+from .state_fidelity_result import StateFidelityResult
+
+
+class BaseStateFidelity(ABC):
+    r"""
+    An interface to calculate state fidelities (state overlaps) for pairs of
+    (parametrized) quantum circuits. The calculation depends on the particular
+    fidelity method implementation, but can be always defined as the state overlap:
+
+    .. math::
+
+            |\langle\psi(x)|\phi(y)\rangle|^2
+
+    where :math:`x` and :math:`y` are optional parametrizations of the
+    states :math:`\psi` and :math:`\phi` prepared by the circuits
+    ``circuit_1`` and ``circuit_2``, respectively.
+
+    """
+
+    def __init__(self) -> None:
+
+        # use cache for preventing unnecessary circuit compositions
+        self._circuit_cache: Mapping[(int, int), QuantumCircuit] = {}
+
+    @staticmethod
+    def _preprocess_values(
+        circuits: QuantumCircuit | Sequence[QuantumCircuit],
+        values: Sequence[float] | Sequence[Sequence[float]] | None = None,
+    ) -> Sequence[Sequence[float]]:
+        """
+        Checks whether the passed values match the shape of the parameters
+        of the corresponding circuits and formats values to 2D list.
+
+        Args:
+            circuits: List of circuits to be checked.
+            values: Parameter values corresponding to the circuits to be checked.
+
+        Returns:
+            A 2D value list if the values match the circuits, or an empty 2D list
+            if values is None.
+
+        Raises:
+            ValueError: if the number of parameter values doesn't match the number of
+                        circuit parameters
+        """
+
+        if isinstance(circuits, QuantumCircuit):
+            circuits = [circuits]
+
+        if values is None:
+            for circuit in circuits:
+                if circuit.num_parameters != 0:
+                    raise ValueError(
+                        f"`values` cannot be `None` because circuit <{circuit.name}> has "
+                        f"{circuit.num_parameters} free parameters."
+                    )
+            return [[]]
+        else:
+            # Support ndarray
+            if isinstance(values, np.ndarray):
+                values = values.tolist()
+            if len(values) > 0 and isinstance(values[0], np.ndarray):
+                values = [v.tolist() for v in values]
+
+            # ensure 2d list
+            if not isinstance(values, list):
```

Comment:
Given we say Sequence, which can be other forms other than list, I think this typecheck needs to be broadened to check against Sequence